sharding/proposer/proposer.go

```diff
@@ -11,12 +11,36 @@ import (
 	"github.com/ethereum/go-ethereum/sharding/mainchain"
 )
 
+// AddHeader adds the collation header to the main chain by sending
+// an addHeader transaction to the sharding manager contract.
+// There can only exist one header per period per shard, it's proposer's
+// responsibility to check if a header has been added.
+func AddHeader(transactor mainchain.ContractTransactor, collation *sharding.Collation) error {
+	log.Info("Adding header to SMC")
+
+	txOps, err := transactor.CreateTXOpts(big.NewInt(0))
+	if err != nil {
+		return fmt.Errorf("unable to initiate add header transaction: %v", err)
+	}
+
+	// TODO: Copy is inefficient here. Let's research how to best convert hash to [32]byte.
+	var chunkRoot [32]byte
+	copy(chunkRoot[:], collation.Header().ChunkRoot().Bytes())
+
+	tx, err := transactor.SMCTransactor().AddHeader(txOps, collation.Header().ShardID(), collation.Header().Period(), chunkRoot)
```

Comment:
I'm really concerned about this pattern. 

This method achieves the following:
- Creates add header transaction
- Broadcasts the transaction to peers
- Prints the pending transaction hash
- Returns nothing

Did the transaction succeed? Did it fail? Was it mined? How would the caller know?

Reply:
Discussed offline. This is being addressed in a different PR. I think #186